File: Medeiros/pmatch.m
function [xyz, db] = pmatch(P1,PR1,P2,PR2,INFO)
% Esse algoritmo calcula as possíveis coordenadas x,y,z considerando que os dois pontos de
entrada P1 e P2 são correspondentes. Também retorna o valor db, que calcula a correspondência
(quanto menor o db, maior a chance de P1 e P2 serem o mesmo ponto no espaço).
% xpr e ypr correspondem à localização do ponto de referência da imagem de que o ponto de
match foi retirado. xci e yci são a localização do centro da imagem.
format compact
xpr1 = PR1(1) ; ypr1 = PR1(2);
xpr2 = PR2(1) ; ypr2 = PR2(2);
xci1 = PR1(3) ; yci1 = PR1(4);
xci2 = PR2(3) ; yci2 = PR2(4);
% As coordenadas de P1 e P2 são traduzidas para ângulos de inclinação com a horizontal
(ANG(1)) e ângulo azimutal (ANG(2)) em radiano.
ANG1d(1) = INFO(1) - 180 * atan((P1(2) - yci1) * INFO(7)/INFO(6))/pi;
ANG1d(2) = INFO(2) - 180 * (atan((P1(1) - xci1) * INFO(7)/INFO(6)) - atan((xpr1 - xci1) *
INFO(7)/INFO(6)))/pi;
ANG2d(1) = INFO(8) - 180 * atan((P2(2) - yci2) * INFO(14)/INFO(13))/pi;
ANG2d(2) = INFO(9) - 180 * (atan((P2(1)-xci2) * INFO(14)/INFO(13)) - atan((xpr2 - xci2) *
INFO(14)/INFO(13)))/pi;
ANG1 = ANG1d * pi/180;
ANG2 = ANG2d * pi/180;
% encontrando a projeção (xp,yp) no plano horizontal para esse par.
xp = -(INFO(11) - INFO(4) - INFO(10) * tan(ANG2(2) - INFO(3) * tan(ANG1(2)))) /
(tan(ANG2(2)) - tan(ANG1(2)));
yp = (xp - INFO(3)) * tan(ANG1(2)) + INFO(4);
% calculando as estimativas de altura para o suposto ponto 3D (x,y,z)
d1 = sqrt((INFO(3) - xp)∧2 + (INF O(4) − yp) ∧ 2);
d2 = sqrt((INFO(10) - xp)∧2 + (INF O(11) − yp) ∧ 2);
zp1 = d1 * tan(ANG1(1)) + INFO(5);
zp2 = d2 * tan(ANG2(1)) + INFO(13);
109
% calculando db de acordo com a diferença entre os valores de zp1 e zp2
db = atan(abs(zp1-zp2)/min(d1,d2));
% retornando a altura estimada como a média dos valores de altura encontrados para cada ponto
z = zp2 + (d2/cos(ANG2(1))) * (zp1-zp2)/(d2/cos(ANG2(1)) + d1/cos(ANG1(1)));
xyz = [xp,yp,z];
end

File: Medeiros/pmatch.py
import numpy as np

def pmatch(P1, PR1, P2, PR2, INFO):
    """
    This algorithm calculates the possible x, y, z coordinates, assuming that the two input points P1 and P2 are corresponding.
    It also returns the value db, which calculates the correspondence (the smaller the db, the higher the chance that P1 and P2 are the same point in space).
    xpr and ypr correspond to the location of the reference point of the image from which the match point was taken. xci and yci are the location of the center of the image.
    """
    # Extract values from PR1 and PR2
    xpr1, ypr1, xci1, yci1 = PR1[0], PR1[1], PR1[2], PR1[3]
    xpr2, ypr2, xci2, yci2 = PR2[0], PR2[1], PR2[2], PR2[3]

    # Convert P1 and P2 coordinates to angles (inclination and azimuthal) in radians
    ANG1d = np.zeros(2)
    ANG1d[0] = INFO[0] - 180 * np.arctan((P1[1] - yci1) * INFO[6] / INFO[5]) / np.pi
    ANG1d[1] = INFO[1] - 180 * (np.arctan((P1[0] - xci1) * INFO[6] / INFO[5]) - np.arctan((xpr1 - xci1) * INFO[6] / INFO[5])) / np.pi

    ANG2d = np.zeros(2)
    ANG2d[0] = INFO[7] - 180 * np.arctan((P2[1] - yci2) * INFO[13] / INFO[12]) / np.pi
    ANG2d[1] = INFO[8] - 180 * (np.arctan((P2[0] - xci2) * INFO[13] / INFO[12]) - np.arctan((xpr2 - xci2) * INFO[13] / INFO[12])) / np.pi

    ANG1 = ANG1d * np.pi / 180
    ANG2 = ANG2d * np.pi / 180

    # Find the projection (xp, yp) on the horizontal plane for this pair
    xp = -(INFO[10] - INFO[3] - INFO[9] * np.tan(ANG2[1] - INFO[2] * np.tan(ANG1[1]))) / (np.tan(ANG2[1]) - np.tan(ANG1[1]))
    yp = (xp - INFO[2]) * np.tan(ANG1[1]) + INFO[3]

    # Calculate height estimates for the supposed 3D point (x, y, z)
    d1 = np.sqrt((INFO[2] - xp)**2 + (INFO[3] - yp)**2)
    d2 = np.sqrt((INFO[9] - xp)**2 + (INFO[10] - yp)**2)

    zp1 = d1 * np.tan(ANG1[0]) + INFO[4]
    zp2 = d2 * np.tan(ANG2[0]) + INFO[12]

    # Calculate db based on the difference between zp1 and zp2
    db = np.arctan(np.abs(zp1 - zp2) / min(d1, d2))

    # Return the estimated height as the average of the height values found for each point
    z = zp2 + (d2 / np.cos(ANG2[0])) * (zp1 - zp2) / (d2 / np.cos(ANG2[0]) + d1 / np.cos(ANG1[0]))
    xyz = np.array([xp, yp, z])

    return xyz, db